pkg/operation/botanist/component/kubecontrollermanager/kube_controller_manager.go

```diff
@@ -621,23 +688,32 @@ func (k *kubeControllerManager) computeCommand(port int32) []string {
 		fmt.Sprintf("--cluster-signing-kube-apiserver-client-key-file=%s/%s", volumeMountPathCAClient, secrets.DataKeyPrivateKeyCA),
 		fmt.Sprintf("--cluster-signing-legacy-unknown-cert-file=%s/%s", volumeMountPathCAClient, secrets.DataKeyCertificateCA),
 		fmt.Sprintf("--cluster-signing-legacy-unknown-key-file=%s/%s", volumeMountPathCAClient, secrets.DataKeyPrivateKeyCA),
+		"--cluster-signing-duration="+pointer.DurationDeref(k.values.ClusterSigningDuration, 720*time.Hour).String(),
+		fmt.Sprintf("--concurrent-endpoint-syncs=%d", pointer.IntDeref(k.values.ControllerWorkers.Endpoint, defaultControllerWorkersEndpoint)),
+		fmt.Sprintf("--concurrent-gc-syncs=%d", pointer.IntDeref(k.values.ControllerWorkers.GarbageCollector, defaultControllerWorkersGarbageCollector)),
+		fmt.Sprintf("--concurrent-service-endpoint-syncs=%d", pointer.IntDeref(k.values.ControllerWorkers.ServiceEndpoint, defaultControllerWorkersServiceEndpoint)),
 	)
 
-	command = append(command,
-		"--cluster-signing-duration=720h",
-		"--concurrent-endpoint-syncs=15",
-		"--concurrent-gc-syncs=30",
-		"--concurrent-namespace-syncs=50",
-		"--concurrent-resource-quota-syncs=15",
-		"--concurrent-service-endpoint-syncs=15",
-		"--concurrent-serviceaccount-token-syncs=15",
-	)
+	if v := pointer.IntDeref(k.values.ControllerWorkers.Namespace, defaultControllerWorkersNamespace); v != 0 {
+		command = append(command, fmt.Sprintf("--concurrent-namespace-syncs=%d", v))
+	}
+
+	if v := pointer.IntDeref(k.values.ControllerWorkers.ResourceQuota, defaultControllerWorkersResourceQuota); v != 0 {
+		command = append(command, fmt.Sprintf("--concurrent-resource-quota-syncs=%d", v))
+		if k.values.ControllerSyncPeriods.ResourceQuota != nil {
+			command = append(command, "--resource-quota-sync-period="+k.values.ControllerSyncPeriods.ResourceQuota.String())
+		}
+	}
+
+	if v := pointer.IntDeref(k.values.ControllerWorkers.ServiceAccountToken, defaultControllerWorkersServiceAccountToken); v != 0 {
+		command = append(command, fmt.Sprintf("--concurrent-serviceaccount-token-syncs=%d", v))
```

Comment:
Comment [here](https://github.com/gardener/gardener/blob/17a3fe5e372f2d1537d34d814e2b5aeff75f8778/pkg/operation/botanist/component/kubecontrollermanager/kube_controller_manager.go#L177-L179) says disabling of controller works only for workerless cluster.
But we are setting concurrent sync here to `0` if the value provided for `ControllerWorkers.ServiceAccountToken` is `0`, independent of whether the cluster has workers or not. Is it not same as disabling the controller because workers count for the controller is set to zero?

Same comment for `ResourceQuota` and `Namespace` controller.

Reply:
> But we are setting concurrent sync here to `0` if the value provided for `ControllerWorkers.ServiceAccountToken` is `0`, independent of whether the cluster has workers or not.

If `v == 0` we don't set `--concurrent-service-acocunt-token-syncs` at all (which basically means that the implicit default of KCM applies). However, this doesn't matter too much. The important part is that we don't add the respective controller to the list of `--controllers` in this case: https://github.com/gardener/gardener/blob/17a3fe5e372f2d1537d34d814e2b5aeff75f8778/pkg/operation/botanist/component/kubecontrollermanager/kube_controller_manager.go#L664-L666

Reply:
Ah Ok, But should we preserve the old behavior for cluster with workers and set the namespace concurrent sync to 50? and same for other controllers. 
But It is fine anyways.
/approve